apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java

```diff
@@ -0,0 +1,236 @@
+  private ManagedChannel channel;
+
+  /**
+   * The primary constructor for this class. Constructs a GrpcRdaSource and opens a channel to the
+   * gRPC service.
+   *
+   * @param config the configuration values used to establish the channel
+   * @param caller the GrpcStreamCaller used to invoke a particular RPC
+   * @param appMetrics the MetricRegistry used to track metrics
+   */
+  public GrpcRdaSource(
+      Config config, GrpcStreamCaller<TResponse> caller, MetricRegistry appMetrics) {
+    this(
+        ManagedChannelBuilder.forAddress(config.host, config.port)
+            .idleTimeout(config.maxIdle.toMillis(), TimeUnit.MILLISECONDS)
+            .build(),
+        caller,
+        appMetrics);
+  }
+
+  /**
+   * This constructor accepts a fully constructed channel instead of a configuration object. This is
+   * used internally by the primary constructor but is also used by unit tests to allow a mock
+   * channel to be provided.
+   *
+   * @param channel channel used to make RPC calls
+   * @param caller the GrpcStreamCaller used to invoke a particular RPC
+   * @param appMetrics the MetricRegistry used to track metrics
+   */
+  @VisibleForTesting
+  GrpcRdaSource(
+      ManagedChannel channel, GrpcStreamCaller<TResponse> caller, MetricRegistry appMetrics) {
+    this.caller = Preconditions.checkNotNull(caller);
+    this.channel = Preconditions.checkNotNull(channel);
+    callsMeter = appMetrics.meter(CALLS_METER);
+    recordsReceivedMeter = appMetrics.meter(RECORDS_RECEIVED_METER);
+    recordsStoredMeter = appMetrics.meter(RECORDS_STORED_METER);
+    batchesMeter = appMetrics.meter(BATCHES_METER);
+  }
+
+  /**
+   * Calls the service through the specific implementation of GrpcStreamCaller provided to our
+   * constructor. Cancels the response stream if reading from the stream is interrupted.
+   *
+   * @param maxPerBatch maximum number of objects to collect into a batch before calling the sink
+   * @param sink to receive batches of objects
+   * @return the number of objects that were successfully processed
+   * @throws ProcessingException wrapper around any Exception thrown by the service or sink
+   */
+  @Override
+  public int retrieveAndProcessObjects(int maxPerBatch, RdaSink<TResponse> sink)
+      throws ProcessingException {
+    callsMeter.mark();
+    boolean interrupted = false;
+    Exception error = null;
+    int processed = 0;
+    try {
+      final GrpcResponseStream<TResponse> responseStream = caller.callService(channel);
+      final List<TResponse> batch = new ArrayList<>();
+      try {
+        while (responseStream.hasNext()) {
+          final TResponse result = responseStream.next();
+          recordsReceivedMeter.mark();
+          batch.add(result);
+          if (batch.size() >= maxPerBatch) {
+            processed += submitBatchToSink(sink, batch);
+          }
+        }
+        if (batch.size() > 0) {
+          processed += submitBatchToSink(sink, batch);
```

Comment:
@brianburton this is a little premature because we're not writing to the DB yet, but is there any chance we could write messages to the DB in a different order in which they arrive (this isn't necessarily a bad thing, but we do need to account for that).  

For example, if RDA had 1001 messages for us to consume and our batch size was 1000.  Say the 1000 record was claim X with a status of suspended and record 1001 was claim X with a status of denied.  Is there any chance that message 1001 will get sent to the sink before 1000?

Reply:
Summary: The plan is for the source to always deliver records to the sink in the order it receives them.

Detail: The current design is sequential over a single thread.  I'm leaving it that way for now (rather than add the complexity of using worker threads for the transformation/writes) until we know what our volumes will be. 

My thoughts for using threads is to add a flush() method to RdaSink and allow the sink to use an internal thread pool for workers that do the data transformation and writing to the database.  The flush() method would be called on pipeline exit to wait for any pending threads to finish writing before shutting down the pipeline.

The sink would still recieve the records in order but might write them out of order.  The progress tracker would have to account for that by tracking what replay tokens have been received but not yet written and then when we resume use the lowest non-written token for the replay parameter.  The sink would safely handle seeing the same records again when necessary (i.e. not creating duplicates or overwriting newer data).

Reply:
Also I plan to store the replay token in the database records so the sink can tell if the database record is newer than the api record it's currently writing.  That should handle the 1001 vs 1000 problem since they would both be the same claim but 1001's replay token would be after 1000's token.